### PR TITLE
UGENE-7010

### PR DESCRIPTION
**Pull request by DmitriiSukhomlinov**
https://ugene.dev/tracker/browse/UGENE-7010

### CODE REVIEW

File: src/plugins/GUITestBase/src/runnables/ugene/plugins/dna_export/ExportMSA2MSADialogFiller.cpp [hunk 1]
```diff
@@ -54,6 +56,9 @@ void ExportMSA2MSADialogFiller::commonScenario() {
         QComboBox *formatCombo = dialog->findChild<QComboBox *>("formatCombo");
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/runnables/ugene/plugins/dna_export/ExportMSA2MSADialogFiller.cpp]: Use GTWidget::findExactWidget. It will assert if the child element is not found.

File: src/plugins/dna_export/src/ExportTasks.h [hunk 2]
```diff
@@ -90,6 +96,7 @@ class ExportMSA2MSATask : public DocumentProviderTask {
     QString url;
     QString format;
     QList<DNATranslation *> aminoTranslations;    // amino translation for a sequences in alignment. If not NULL -> sequence is translated
+    bool trimGaps;
```
Comment (yalgaer) [inline src/plugins/dna_export/src/ExportTasks.h]: Please document any new methods/fields added.

File: src/plugins/GUITestBase/src/tests/common_scenarios/msa_editor/GTTestsMsaEditor.cpp [hunk 3]
```diff
@@ -4615,6 +4615,32 @@ GUI_TEST_CLASS_DEFINITION(test_0090) {
...
+    // 4. Copy to clipboard
+    GTUtilsMSAEditorSequenceArea::selectArea(os);
+    GTUtilsDialog::waitForDialog(os, new PopupChooser(os, QStringList() << "MSAE_MENU_COPY"
+                                                                        << "copy_selection"));
+    GTMouseDriver::click(Qt::RightButton);
+    GTUtilsTaskTreeView::waitTaskFinished(os);
+
+    // Expected: TAVS\nXXVS
+    const QString clipboardText = GTClipboard::sequences(os);
+    const QString expectedMSA = "TAVS\nXXVS";
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/tests/common_scenarios/msa_editor/GTTestsMsaEditor.cpp:4640]: Why have you selected 'X' option instead of '-' variant? 
I do not know which is better here and unless you have a strong opinion that 'X' is better than '-' I would leave this decision to a user: ask user if to translate incomplete triplets to 'X' or to '-'.

One of the sources for a 'strong opinion' would be a reference to some existing tool that prefers one over another.

File: src/plugins/GUITestBase/src/tests/common_scenarios/msa_editor/GTTestsMsaEditor.cpp [hunk 4]
```diff
@@ -4615,6 +4615,32 @@ GUI_TEST_CLASS_DEFINITION(test_0090) {
...
+
+    // 2. Do document context menu {Export->Export aligniment to amino format}
+    // 3. Translate with "Include gaps"
+    GTUtilsDialog::waitForDialog(os, new ExportMSA2MSADialogFiller(os, -1, sandBoxDir + "GUITest_common_scenarios_msa_editor_test_0091.aln", true));
+    GTUtilsDialog::waitForDialog(os, new PopupChooser(os, QStringList() << MSAE_MENU_EXPORT << "amino_translation_of_alignment_rows"));
+    GTWidget::click(os, GTUtilsMdi::activeWindow(os), Qt::RightButton);
+    GTUtilsTaskTreeView::waitTaskFinished(os);
+
+    // 4. Copy to clipboard
+    GTUtilsMSAEditorSequenceArea::selectArea(os);
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/tests/common_scenarios/msa_editor/GTTestsMsaEditor.cpp:4632]: nit: this normally should be a 1-liner, like getWholeData()..

**Review by yalgaer — CHANGES_REQUESTED** [review]
Everything looks good, except the final solution remains unclear: why 'X' over  '-' or the opposite.

If we have no good examples in other tools I would suggest to stay non opinionated and allow a user to make the decision.

**Comment by yalgaer** [comment]
++ Could you please add everyone from the team to the list of the reviewers by default.

File: src/plugins/GUITestBase/src/runnables/ugene/plugins/dna_export/ExportMSA2MSADialogFiller.cpp (hunk 1 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/GUITestBase/src/runnables/ugene/plugins/dna_export/ExportMSA2MSADialogFiller.cpp]: Done

File: src/plugins/GUITestBase/src/tests/common_scenarios/msa_editor/GTTestsMsaEditor.cpp (hunk 4 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/GUITestBase/src/tests/common_scenarios/msa_editor/GTTestsMsaEditor.cpp:4632]: Sorry, don't get the idea

File: src/plugins/GUITestBase/src/tests/common_scenarios/msa_editor/GTTestsMsaEditor.cpp (hunk 3 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/GUITestBase/src/tests/common_scenarios/msa_editor/GTTestsMsaEditor.cpp:4640]: I have added a possibility to choose

**Comment by DmitriiSukhomlinov** [comment]
Have fixed all comments, please, review again.

File: src/plugins/GUITestBase/src/tests/common_scenarios/msa_editor/GTTestsMsaEditor.cpp (hunk 4 above)
Comment (yalgaer) [reply src/plugins/GUITestBase/src/tests/common_scenarios/msa_editor/GTTestsMsaEditor.cpp:4632]: If you want to check the sequence area content you can use a single line to access it: getWholeData. No need to copy/check clipboard for that. That will keep the test code minimal and focused in the key subject.
This is not a blocker anyway - only a suggestion.

File: src/plugins/dna_export/src/ExportMSA2MSADialog.h [hunk 5]
```diff
@@ -35,9 +35,16 @@ class ExportMSA2MSADialog : public QDialog, Ui_ExportMSA2MSADialog {
 public:
     ExportMSA2MSADialog(const QString &defaultFileName, const DocumentFormatId &f, bool wholeAlignmentOnly, QWidget *p);
 
+    enum class UnknownAmino {
```
Comment (yalgaer) [inline src/plugins/dna_export/src/ExportMSA2MSADialog.h:42]: Please document any new types/methods/fields.

I would suggest to use "above the subject" doc style like /** some text */  because it will never go to the line below and automatically shown by IDEs. 

The suffix-style comments like 
int fieldX; // fieldX is lorem ipsum...
are good only if they are short (single line) and are never shown by IDE as contextual docs. I think they are good for the inner comments, but not for the top-level once.

File: src/plugins/dna_export/src/ExportMSA2MSADialog.ui [hunk 6]
```diff
@@ -61,6 +61,51 @@
...
+     </property>
+    </widget>
+   </item>
+   <item>
+    <widget class="QGroupBox" name="gbTranslateAs">
+     <property name="enabled">
+      <bool>false</bool>
+     </property>
+     <property name="title">
+      <string>Translate unknown amino acid as:</string>
```
Comment (yalgaer) [inline src/plugins/dna_export/src/ExportMSA2MSADialog.ui]: I would not say "translate amino acid" because you translate "nucleic acids" here.
I would rephrase it to "Translate triplets with gaps as:"

File: src/plugins/dna_export/src/ExportTasks.cpp [hunk 7]
```diff
@@ -179,6 +193,9 @@ void ExportMSA2MSATask::run() {
             aminoTT->translate(seq.constData(), seq.length(), resseq.data(), resseq.length());
 
             resseq.replace("*", "X");
+            if (!trimGaps && convertUnknownToGap) {
+                resseq.replace("X", "-");
```
Comment (yalgaer) [inline src/plugins/dna_export/src/ExportTasks.cpp:196]: This is still not correct. 

Here is 2 reasons why:
1) We can't replace all X with gaps, because you replace stop codons (valid nucleic triplets, non-gaps) with gaps.
The simplest solution I see here is to move "resseq.replace("*", "X");" below your replacement code

2) There is a chance that some incomplete DNA triplets are always converted to the same amino. 
For example: Proline, Serine, etc... may have "any" base in the end position and still be translated into a valid protein.
https://miro.medium.com/max/1440/0*YVhxcZ55DqPBDdGF.png

This way if in the original DNA alignment 'gap' means 'unknown', the translation still should contain a valid amino acid, not a gap. 

I'm OK if the item 2) is fixed as a separate bug/PR.

File: src/plugins/dna_export/src/ExportTasks.h [hunk 8]
```diff
@@ -90,6 +97,8 @@ class ExportMSA2MSATask : public DocumentProviderTask {
     QString url;
     QString format;
     QList<DNATranslation *> aminoTranslations;    // amino translation for a sequences in alignment. If not NULL -> sequence is translated
+    bool trimGaps;                                // trim gaps before translation of not
```
Comment (yalgaer) [inline src/plugins/dna_export/src/ExportTasks.h]: nit: if something is not expected to be changed inside of the class it can be declared as a 'const'

**Review by yalgaer — CHANGES_REQUESTED** [review]
The last, but not the least: could you please run Linux Nighly for the PR, so we know that it does not break any existing behavior and the tests from the PR are green?

File: src/plugins/dna_export/src/ExportMSA2MSADialog.h (hunk 5 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/dna_export/src/ExportMSA2MSADialog.h:42]: Done

File: src/plugins/dna_export/src/ExportMSA2MSADialog.ui (hunk 6 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/dna_export/src/ExportMSA2MSADialog.ui]: Done

File: src/plugins/dna_export/src/ExportTasks.cpp (hunk 7 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/dna_export/src/ExportTasks.cpp:196]: 1. Fixed
2. I understand the idea, but, actually, don't get a sequence, which could be a part of alignment and affected by this problem. Could you help me?

File: src/plugins/dna_export/src/ExportTasks.h (hunk 8 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/dna_export/src/ExportTasks.h]: Done

File: src/plugins/GUITestBase/src/tests/common_scenarios/msa_editor/GTTestsMsaEditor.cpp (hunk 4 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/GUITestBase/src/tests/common_scenarios/msa_editor/GTTestsMsaEditor.cpp:4632]: I took a note, but didn't fix it this time

**Comment by DmitriiSukhomlinov** [comment]
@yalgaer fixed, check again

File: src/plugins/dna_export/src/ExportTasks.cpp (hunk 7 above)
Comment (yalgaer) [reply src/plugins/dna_export/src/ExportTasks.cpp:196]: Ok, I will create a separate issue.

**Comment by yalgaer** [comment]
The PR is ready for the approval. No code related issues left. Just run the Linux/Nightly in the Teamcity for your branch to show no regressions added by this commit.
You can put your build in front of the automatically scheduled builds.

**Comment by DmitriiSukhomlinov** [comment]
Have passed [here](http://ugene-teamcity.unipro.ru:8111/ugene-teamcity/viewType.html?buildTypeId=bt153&branch_UgeneNightlyBuilds_GuiTests=UGENE-7010&tab=buildTypeStatusDiv). @yalgaer, approve, please.

**Review by yalgaer — APPROVED** [review]
